aten/src/ATen/core/DistributionsHelper.h

```diff
@@ -0,0 +1,288 @@
+ *      auto sample = uniform(gen);
+ */
+
+
+namespace at {
+
+// Using VectorType in Box-muller derived distributions to avoid
+// code duplication
+template <typename T>
+struct VectorType {  };
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <> struct VectorType<half> { using type = at::detail::Array<float, 2>; };
+#endif
+template <> struct VectorType<Half> { using type = at::detail::Array<float, 2>; };
+template <> struct VectorType<float> { using type = at::detail::Array<float, 2>; };
+template <> struct VectorType<double> { using type = at::detail::Array<double, 2>; };
+
+template <typename T>
+using vect_type = typename VectorType<T>::type;
+
+// Using DistAccumType in accumulate types for distributions.
+// Note: Ideally we'd be using ATen/AccumulateType.h but looks
+// like the there is some inconsistency in how accumulate types
+// are mapped currently, e.g. for the cpu side, float is mapped
+// to double.
+template <typename T>
+struct DistAccumType {  };
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+template <> struct DistAccumType<half> { using type = float; };
+#endif
+template <> struct DistAccumType<Half> { using type = float; };
+template <> struct DistAccumType<float> { using type = float; };
+template <> struct DistAccumType<double> { using type = double; };
+
+template <typename T>
+using dist_acctype = typename DistAccumType<T>::type;
+
+// Constants for uniform distribution
+// doubles have 52 bits of mantissa (fractional part)
+constexpr uint64_t DOUBLE_MASK = (1ULL << 53) - 1;
+constexpr double DOUBLE_DIVISOR = 1.0 / (1ULL << 53);
+
+// floats have 23 bits of mantissa (fractional part)
+constexpr uint32_t FLOAT_MASK = (1 << 24) - 1;
+constexpr float FLOAT_DIVISOR = 1.0f / (1 << 24);
+
+/**
+ * Samples a uniform distribution in the range [0,1) of type T
+ */
+template <typename T>
+struct uniform_real_distribution {
+
+  C10_HOST_DEVICE inline uniform_real_distribution(T a_in, T b_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(a_in <= b_in);
+        assert(b_in-a_in <= std::numeric_limits<T>::max());
+      #else
+        AT_ASSERT(a_in <= b_in);
+        AT_ASSERT(b_in-a_in <= std::numeric_limits<T>::max());
+      #endif
+    #endif
+    a = a_in;
+    b = b_in;
+  }
+
+  C10_HOST inline dist_acctype<T> operator()(at::CPUGenerator* generator){
+    dist_acctype<T> x;
+    if(std::is_same<T, double>::value) {
+      x = (generator->random64() & DOUBLE_MASK) * DOUBLE_DIVISOR;
+    } else {
+      x = (generator->random() & FLOAT_MASK) * FLOAT_DIVISOR;
+    }
+    return (x * (b - a) + a);
+  }
+
+  private:
+    T a;
+    T b; 
+};
+
+/**
+ * Samples a normal distribution using the Box-Muller method
+ * Takes mean and standard deviation as inputs
+ */
+template <typename T>
+struct normal_distribution {
+
+  C10_HOST_DEVICE inline normal_distribution(T mean_in, T stdv_in) {
+    #ifndef __HIP_PLATFORM_HCC__
+      #ifdef __CUDA_ARCH__
+        assert(stdv_in > 0);
```

Comment:
Ditto here. The reason we prefer not to use `assert` is because, in optimization mode, these get compiled away. But you really do want to always check if the `stdv` is nonsense, since it might be user-supplied.